usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -489,6 +491,394 @@ static CK_RV cleanse_attribute(TEMPLATE *template,
+        return rc;
+    }
+    if (abwrapkey) {
+        if (!abkey) {
+            TRACE_ERROR("Wrapping key is AB, but target key not\n");
+            return CKR_KEY_NOT_WRAPPABLE;
+        }
+        /* Here, only AB wrapping can be used.  Check mechanism and parameters. */
+        if (mech->mechanism != CKM_IBM_ATTRIBUTEBOUND_WRAP ||
+            sizeof(CK_IBM_ATTRIBUTEBOUND_WRAP_PARAMS) != mech->ulParameterLen) {
+            TRACE_ERROR("AB key wrapping attempt without CKM_IBM_ATTRIBUTEBOUND_WRAP");
+            return CKR_MECHANISM_INVALID;
+        }
+        rc = check_ab_kek_type(wrap_key_obj->template, CKR_WRAPPING_KEY_TYPE_INCONSISTENT);
+        if (rc != CKR_OK)
+            return rc;
+        rc = h_opaque_2_blob(tokdata,
+                             ((CK_IBM_ATTRIBUTEBOUND_WRAP_PARAMS *)mech->pParameter)->hSignVerifyKey,
+                             sblob, sblob_len, sobj, READ_LOCK);
+        if (rc != CKR_OK) {
+            TRACE_ERROR("Failed to get sign key for AB wrapping\n");
+            return rc;
+        }
+        rc = template_attribute_get_bool((*sobj)->template, CKA_IBM_ATTRBOUND, &absignkey);
+        if (rc != CKR_OK || !absignkey) {
+            TRACE_ERROR("AB-Wrap: sign key not AB\n");
+            rc = CKR_KEY_FUNCTION_NOT_PERMITTED;
+            goto out;
+        }
+        rc = template_attribute_get_bool((*sobj)->template, CKA_SIGN, &signsignkey);
+        if (rc != CKR_OK || !signsignkey) {
+            TRACE_ERROR("AB-Wrap: sign key not able to sign\n");
+            rc = CKR_KEY_FUNCTION_NOT_PERMITTED;
+            goto out;
+        }
+    } else if (abkey) {
+        TRACE_ERROR("Target key is AB, but wrapping key is not\n");
+        return CKR_KEY_FUNCTION_NOT_PERMITTED;
+    }
+ out:
+    if (rc != CKR_OK && *sobj) {
+        object_put(tokdata, *sobj, TRUE);
+        *sobj = 0;
+        *sblob = 0;
+        *sblob_len = ~0;
+    }
+    return rc;
+}
+
+static CK_RV check_ab_unwrap(STDLL_TokData_t * tokdata,
+                             CK_BYTE **vblob, size_t *vblob_len,
+                             OBJECT **vobj, OBJECT *unwrap_key_obj,
+                             CK_ATTRIBUTE_PTR attrs, CK_ULONG attrs_len,
+                             CK_MECHANISM_PTR mech, CK_BBOOL *isab)
+{
+    CK_RV rc;
+    CK_BBOOL abkey, abunwrapkey, abverifykey, verifyverifykey;
+
+    *vobj = 0;
+    *vblob = 0;
+    *vblob_len = ~0;
+    *isab = FALSE;
+    rc = template_attribute_get_bool(unwrap_key_obj->template, CKA_IBM_ATTRBOUND,
+                                     &abunwrapkey);
+    if (rc == CKR_TEMPLATE_INCOMPLETE)
+        abunwrapkey = FALSE;
+    else if (rc != CKR_OK)
+        return rc;
+    rc = get_bool_attribute_by_type(attrs, attrs_len, CKA_IBM_ATTRBOUND, &abkey);
+    if (rc == CKR_TEMPLATE_INCOMPLETE)
+        abkey = FALSE;
+    else if (rc != CKR_OK)
+        return rc;
+    if (mech->mechanism == CKM_IBM_ATTRIBUTEBOUND_WRAP) {
```

Comment:
If we would check for (mech->mechanism != CKM_IBM_ATTRIBUTEBOUND_WRAP) and do the cleanup at line 715 here, we would save brackets and indentation ... just for better reading.

Reply:
There is an `else` attached to this `if` capturing the cases where someone wants to perform AB-Wrapping without AB Keys.  So even if I invert the logic, I still need to do all the code in the else-branch here.  This sill not save any brackets or indentation.